gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -188,270 +195,168 @@ public Object call() throws IOException {
-        try {
-          responseObserver.done.await();
-        } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-          throw new IOException(
-              String.format("Resumable upload failed for '%s'", getResourceString()), e);
+      long writeOffset = retriesAttempted > 0 ? getCommittedWriteSize(uploadId) : 0;
+      objectHasher = retriesAttempted > 0 ? objectHasher : Hashing.crc32c().newHasher();
+
+      responseObserver = new InsertChunkResponseObserver(uploadId, writeOffset);
+      // TODO(b/151184800): Implement per-message timeout, in addition to stream timeout.
+      StreamObserver<InsertObjectRequest> requestStreamObserver =
+          stub.withDeadlineAfter(WRITE_STREAM_TIMEOUT.toMillis(), MILLISECONDS)
+              .insertObject(responseObserver);
+
+      boolean objectFinalized = false;
+      while (!objectFinalized) {
+        InsertObjectRequest insertRequest;
+        if (dataChunkMap.size() > 0 && dataChunkMap.lastKey() >= writeOffset) {
+          insertRequest = buildRequestFromBufferedDataChunk(dataChunkMap, writeOffset);
+          writeOffset += insertRequest.getChecksummedData().getContent().size();
+        } else {
+          ByteString data =
+              ByteString.readFrom(ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE));
+          dataChunkMap.put(writeOffset, data);
+          if (dataChunkMap.size() >= NUMBER_OF_REQUESTS_TO_RETAIN) {
+            dataChunkMap.remove(dataChunkMap.firstKey());
+          }
+          insertRequest = buildInsertRequest(writeOffset, data, false);
+          writeOffset += data.size();
         }
+        requestStreamObserver.onNext(insertRequest);
+        objectFinalized = insertRequest.getFinishWrite();
 
         if (responseObserver.hasTransientError()) {
-          writeOffset = getCommittedWriteSize(uploadId);
-          ++retriesAttempted;
-        } else {
-          writeOffset += responseObserver.bytesWritten();
-          retriesAttempted = 0;
+          retriesAttempted++;
+          return doResumableUpload();
         }
+      }
 
-        if (retriesAttempted >= UPLOAD_RETRIES) {
-          throw new IOException(
-              String.format(
-                  "Too many retry attempts. Resumable upload failed for '%s'", resourceId));
+      requestStreamObserver.onCompleted();
+
+      try {
+        responseObserver.done.await();
+        if (responseObserver.hasTransientError()) {
+          retriesAttempted++;
+          return doResumableUpload();
         }
-      } while (!responseObserver.hasFinalized());
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+        throw new IOException(
+            String.format("Resumable upload failed for '%s'", getResourceString()), e);
+      }
 
       return responseObserver.getResponseOrThrow();
     }
 
+    private InsertObjectRequest buildInsertRequest(
+        long writeOffset, ByteString dataChunk, boolean resumeMode) {
```

Comment:
What does `resumeMode` means? Is this mean recovery of failed upload? If so, let's chose better name to reflect this.

Reply:
Renamed resumeFromFailedInsert to better reflect the case.